### ISSUE TITLE
House_price_prediction

### ISSUE THREAD

**Issue by HariKrishna1-DS** [HariKrishna1-DS/NLP#1]

**Comment by HariKrishna1-DS** [comment]
import pandas as pd
data = pd.read_csv("C:\\Users\\sriha\\OneDrive\\Desktop\\New folder\\pb excel\\Bengaluru_House_Data.csv")
df = pd.DataFrame(data)
print (df.head())
print (df.area_type. unique())
print (df.groupby('area_type')['area_type'].agg('count'))
print (df.isnull().sum())
df1 =df.drop (['availability','society','balcony','area_type'], axis ="columns")
print (df1.head())
# finding null values
print (df1.isnull().sum())
'''' there is null values are 90 
    remove that null values
'''

df2= df1.dropna()
#print (df2.isnull().sum())
#print (df2.shape)       # 13320 - 13246 = 74
#print (df2.head())
#print( df2.location.unique())
#print(len(df2.location.unique()))
#print(df2['size'].unique())
''' Now here is problem that
    that, I want to separate the BHK numbers
                    separate the BHK charcters'''

df2['BHK'] = df2['size'].apply (lambda x : int(x.split(' ')[0]))
#print(df2.head())
#print (df2.BHK.unique())   # there are number of BHK houses are there 
# now lets check above BHK> 20 
print(df2[df2.BHK> 20])
#print(df2.total_sqft.unique())





''' Now here a problem that all sqft_values are numbers BUT,
                   some are     sqft_values are IN BETWEEN values
                   
    Let find how many there are'''

df3= df2.copy()
#print(df3.head())
df3.drop('size', axis ="columns", inplace=True)
#print(df3.head())
# let check how many values are '1133 - 1384' like between values 
def btwn_values(x):
    try: 
        float(x)
    except:
        return False
    return True

#print(df3[~(df3.total_sqft.apply (btwn_values))])




# there are 190 values there 
'''So, that values I dont want to remove Instead of
   Give Average for that values like
   (2100 + 2850)/ 2 ===> 2950/ 2====> 2475
   '''
   #  Now these values to convert into AVERAGE values 
#  Like (x+y)/2 


def convert_btwn_values(x):
    splits = x.split('-')
    if len (splits)==2:
        return (float (splits[0]) + float(splits[1])) / 2
    try:
        return float(x)
    except :
        return None
    
#print(convert_btwn_values('1200 - 1470'))    # 2670/2  ====> 1335
#print(convert_btwn_values('1056'))



# Now I want  these average  values into  main data
df4 = df3.copy()
df4.total_sqft = df4['total_sqft'] .apply(convert_btwn_values)
print (df4.total_sqft.unique())
#print(df4.loc[122])
#print (df4.head())



# 2. FUTURE ENGINEERING



'''1. To find price_Per_sqft.
    2. Location no:of places '''
    
#print(df4.head())
#print (df4.bath.unique())
#print(df4[df4.bath> 10])
print(len(df4[df4.bath>=10]))

#  To find the Number of Location

#print (len(df4.location.unique()))


df4.location = df4.location.apply(lambda x : x.strip())
location_status = df4.groupby('location')['location'].agg('count').sort_values(ascending =False)
#print (location_status)
#print(location_status[location_status<=10])


# Now I gonna to create all which is less 10 places that  named as "others"
location_less_than_10 = location_status[location_status<=10]
#print(location_less_than_10)


df4.location = df4.location.apply (lambda x : "OTHER <=10 " if x in location_less_than_10 else x)
#print(df4.head(20))

# Here Comes the houses and apartments are ehich is less than 10 floors apartment location
# are written as '1-10 floors apartment'

'''Lets find out the Price_per_sqft  for total_sqft'''

df5 = df4.copy()
#print(df5.head())

df5['price_per_sqft'] = df5['price'] * 100000/ df5['total_sqft'] 
#print(df5.head())

# print(df5.shape)









# 3. OUTLIERS REMOVAL
#   1. FIND total_sqft/ BHKs  per BHK 
#   2. REMOVE THE LESS AND HIGH MEAN VALUES OF  MEAN AND STD
df6 = df5.copy()
# print (df6.head())

df6['total_sqft/BHKs'] = df6['total_sqft'] / df6['BHK']
#print(df6.head())


'''The term "1BHK" or "one-bedroom house" refers to a home with
just one bedroom, one hall, one kitchen, one bathroom, and one dining/living room, 
which together take up between 450 and 600 square feet of carpet space'''

# That way I want to remove which is less than total_sqft/BHK <=300
df6[df6['total_sqft/BHKs'] <=300]
#The lookout type of table and there are nearly 936 values are there,
# so I want to remove that  which is <= 300 from my table..

df7= df6[~(df6['total_sqft/BHKs']<= 300)]
#print(df7.shape)
df6.shape
#13246-12320

(df7.head())


df8= df7.copy()
df8.drop('total_sqft/BHKs', axis="columns", inplace=True)
#print(df8.head())

#print (df8.price_per_sqft.describe())


#  This important in Price_prediction



# According to Middle class they cam't buy 'Low' and 'Very Much High' , so all are looking middle range of rents 
# That way I want to remove all  'min' and 'Max' values.
#  so, I want  to remove the all less  MEAN and SD values.
import pandas as pd
def  outlier_removal(df):
    df_out = pd.DataFrame()
    for keys , sub_df in df.groupby('location'):
        
        mean = np.mean(sub_df.price_per_sqft)
        std  = np.std (sub_df.price_per_sqft)
        
        removal_df =sub_df[(sub_df.price_per_sqft > (mean-std )) & (sub_df.price_per_sqft <= (mean+std))]
        df_out = pd.concat([removal_df, df_out] ,ignore_index= True)
        
    return df_out


print (df8.shape)  # Before removing
import numpy as np
df9 = outlier_removal(df8)  # after removing
#print(df9.shape)


'''
There are nearly 2000 values are removed.....
now when compared to same BHK House size is same but, the Price is also a different example --> uttarhalli location ,the total_sqft of house is 2BHK pricer is 81,000 same location the total_sqft of house is 2BHK pricer is 1,21,000

        The price is totaly different 
        
        
        Like that we find like similar cases in datasets
        
        
        By using Scatter diagrams we find what are simliar BHKs and Prices are the different'''
        
        
#print(df9.head())

(df9.location.head())
#print(df9.location.unique(), "\n", len(df9.location.unique()))



# By we see the 3BHK and 2BHK houses are using Visual representation







import matplotlib.pyplot as pp
def scatter_plot(df,location):
    bhk_2 = df[(df.location==location) & (df.BHK==2)]
    bhk_3 = df[(df.location==location) & (df.BHK==3)]
    pp.scatter (bhk_2.total_sqft, bhk_2.price , c="red", label= "BHK_2")
    pp.scatter (bhk_3.total_sqft, bhk_3.price , c="green", label = "BHK_3")
    pp.legend()
    pp.xlabel('total_sqft')
    pp.ylabel('price')
    pp.grid()
    pp.show()
    
scatter_plot(df9, "Uttarahalli")






'''So when I'm looking the image so it comes 1300 squared feet area
    that it comes  2_bhk house and 3_BHK House
    At same price .....
    So, like that that prices I want to remove that from my data.'''
    
def remove_bhk_outliers(df):
    which_combined = np.array([ ])
    
    for location, location_df in df.groupby('location'):
        bhk_stats ={}
        
        for BHK, BHK_df  in location_df.groupby('BHK'):
            bhk_stats[BHK]={
                'mean' : np.mean(BHK_df['price_per_sqft']),
                'std'  : np.std (BHK_df['price_per_sqft']),
                'count': BHK_df.shape[0]
            }
            
        for BHK, BHK_df in location_df.groupby('BHK'):
            stats = bhk_stats.get(BHK -1)
            if stats and stats['count'] >5 :
                which_combined =np.append(which_combined,BHK_df[BHK_df['price_per_sqft']< (stats['mean'])]. index.values)
                
    return df.drop (which_combined, axis= 'index')
#print(df9.shape)

df10=remove_bhk_outliers(df9)    # There are Nearly 3000  like are removed 
#print(df10.shape)


#print(scatter_plot(df9,"Uttarahalli"))

#print(df10.shape)


'''Now just look at the  how much Squarefeet are high selling  in the all cities.'''




pp.hist(df10.price_per_sqft ,rwidth=0.6)
pp.xlabel("total_sqft_in_Banglore")
pp.ylabel("number_of_counts")
pp.grid()
#pp.show()


#print (df10.bath.unique())

'''So now let's look at that bathroom.
# The number of BHK == bathrooms , if it considered
# The number of BHK == bathrooms + 2 , remove that .....
'''

#print(df10[df10.bath > df10.BHK+2 ])

df11 = df10[df10.bath < df10.BHK+2 ]   # nearly 100 are removed
#print(df11.shape)


# here now data cleaning and everthing over 
# next performing using Machine Learning 
# Before we removing Unnessary columns


#print (df11.head())
df12 = df11.drop('price_per_sqft' , axis="columns")
#print(df12.head())



# 4. MODEL BUILDING ----> Using Machine learning

#print(df12.shape)

#print(len(df12.location.unique()))


dummy  = pd.get_dummies(df12.location)
(dummy.head())
(dummy.shape)

df13 = pd.concat([df12,dummy], axis ="columns")
df13.head()

df14 = df13.drop("location", axis ="columns")
df14.head()
(df14.shape)


#                               Y= mX + C

X = df14.drop('price', axis ="columns")  # Dependent variable
y = df14.price            # Independent Variable

(X.head() ," \n" , y.head())



from sklearn.model_selection import train_test_split
X_train, X_test, y_train ,y_test = train_test_split (X,y, test_size=0.2)

(len(X_train) , len (X_test), len (X))




from sklearn. linear_model import LinearRegression
Lr = LinearRegression()
Lr.fit(X_train,y_train) 
(Lr.score(X_test,y_test))

y_pred= Lr.predict(X_test)
(y_pred)






'''ShuffleSplit = It is used for cross-validation, which is a technique for 
                    assessing the performance of a model by dividing the dataset into different training and testing sets.
                    It uses to Random Shuffle , Create Multiple Splits , Adjustable parameters'''

from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import cross_val_score

cv =ShuffleSplit(n_splits=10, test_size=0.2)
(cross_val_score(LinearRegression(),X,y, cv=cv))



from sklearn.tree import DecisionTreeRegressor
DTR = DecisionTreeRegressor()
DTR.fit(X_train,y_train)
(DTR.score(X_test,y_test))


(X.head())
(X.columns)



#predict the Price on location

def predicted_price(location,total_sqft,bath,BHK):
    location_index = np.where(X.columns ==location)[0][0]
    
    x  = np.zeros(len(X.columns))
    x[0]= total_sqft
    x[1]= bath
    x[2]= BHK
    
    if location_index >=0:
        x[location_index] =1
        
        
    return Lr.predict([x])[0]

print(predicted_price("Uttarahalli", 1800,2,2))
print(predicted_price("Uttarahalli",1600,2,2))

``
